File: ScriptGraph/BuiltInJS/CGVariableNode.js
/**
 * @file CGVariableNode.js
 * @author
 * @date 2021/8/24
 * @brief CGVariableNode.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const Amaz = effect.Amaz;
const {BaseNode} = require('./BaseNode');

class CGVariableNode extends BaseNode {
  constructor() {
    super();
    this.variables = null;
    this.variableName = null;
    this.accessType = null;
  }

  getOutput(index) {
    return this.variables[this.variableName];
  }

  execute(index) {
    if (this.accessType === 'SETTER') {
      let inputValue = this.inputs[1]();
      if (inputValue === null) {
        if (this.valueType === 'Texture2D') {
          inputValue = this.resource;
        }
      }
      this.variables[this.variableName] = inputValue;
      if (this.nexts[0]) {
        this.nexts[0]();
      }
    }
  }
}

exports.CGVariableNode = CGVariableNode;

File: ScriptGraph/BuiltInLua/CGVariableNode.lua
local CGVariableNode = CGVariableNode or {}
CGVariableNode.__index = CGVariableNode

function CGVariableNode.new()
  local self = setmetatable({}, CGVariableNode)
  self.inputs = {}
  self.nexts = {}
  self.variables = nil
  self.variableName = nil
  self.accessType = nil
  return self
end

function CGVariableNode:setInput(index, func)
  self.inputs[index] = func
end

function CGVariableNode:setNext(index, func)
  self.nexts[index] = func
end

function CGVariableNode:getOutput(index)
  return self.variables[self.variableName]
end

function CGVariableNode:execute()
  if self.accessType == 'SETTER' then
    local inputValue = self.inputs[1]()
    if inputValue == nil then
      if self.valueType == 'Texture2D' then
        inputValue = this.resource
      end
    end
    self.variables[self.variableName] = inputValue
    if self.nexts[0] ~= nil then
      self.nexts[0]()
    end
  end
end

return CGVariableNode